src/main/scala/com/datastax/gatling/plugin/model/DseCqlStatements.scala

```diff
@@ -39,33 +38,49 @@ case class DseCqlSimpleStatement(statement: SimpleStatement) extends DseCqlState
   *
   * @param preparedStatement the prepared statement on which to bind parameters
   */
-case class DseCqlBoundStatementNamed(cqlTypes: CqlPreparedStatementUtil, preparedStatement: PreparedStatement)
-    extends DseCqlStatement {
+case class DseCqlBoundStatementNamed(cqlTypes: CqlPreparedStatementUtil,
+                                     preparedStatement: PreparedStatement,
+                                     builderFn: (BoundStatement) => BoundStatementBuilder)
```

Comment:
I am not quite sure I understand why the new `builderFn` parameter is for.  Could you describe its utility?

Reply:
As far as I can tell, it seems that it is a way to avoid calling `new BoundStatementBuilder(...)` when running tests, because otherwise many getters would need to be mock'd.  If so, then there might be a better form, following the same design.

Reply:
Here is an alternative design.

You could first create a companion object for `DseCqlStatement` that would contain the default builder as follows:

```scala
object DseCqlStatement {
  private[model] val defaultBuilder: BoundStatement => BoundStatementBuilder = new BoundStatementBuilder(_)
}
```

Then, import it so that you have a shortcut to reference it.  And everywhere you used to request a default builder, you now add a default value for that parameter.  For example, in `DseCqlBoundStatementName`, it looks like this:

```scala
import com.datastax.gatling.plugin.model.DseCqlStatement.defaultBuilder

// [...]

case class DseCqlBoundStatementNamed(cqlTypes: CqlPreparedStatementUtil,
                                     preparedStatement: PreparedStatement,
                                     builderFn: BoundStatement => BoundStatementBuilder = defaultBuilder)
  extends DseCqlStatement[BoundStatement, BoundStatementBuilder] {
// [...]
}
```

That should allow you to remove a lot of companion objects that you created to hold these default builders.  The test code stays the same, in that it can override the builder function when needed.

The only gotcha will happen in `DseCqlBoundStatementWithPassedParams`.  It currently uses a varargs of `Expression[AnyRef]`, which cannot be used with default parameter values.  So you will need to change it so that it becomes a `Seq[Expression[AnyRef]]`.  With the optional builder parameter being last, it will look like this:

```scala
case class DseCqlBoundStatementWithPassedParams(cqlTypes: CqlPreparedStatementUtil,
                                                preparedStatement: PreparedStatement,
                                                params: Seq[Expression[AnyRef]],
                                                builderFn: BoundStatement => BoundStatementBuilder = defaultBuilder)
// [...]
```

Reply:
Your intuition is entirely correct @pingtimeout .  My original version of this code looked very much like your proposed solution but the issue with DseCqlBoundStatementWithPassedParams made me nervous; I didn't want to introduce yet more API changes as part of this PR.  So I settled on the version implemented here, which is really just an extra constructor on the case class.

I still don't _love_ the idea of converting the params in DseCqlBoundStatementWIthPassedParams into a Seq... I kinda like the vararg behaviour you have now.  Fortunately, I think I can also solve the problem (and remove all the companion objects) via implicit args.  Take a look at the new version and let me know what you think!